Documentation/design-docs/code-generation-strategies.md

```diff
@@ -0,0 +1,37 @@
+# Code generation and execution strategies in CoreCLR
+
+When a .NET language compiler (such as C#, VB.net, or F#) compiles source code into a binary form, the source of the program gets expressed as a sequence of instructions for an abstract processor ("IL instructions"). The abstract processor deals with the program's code at a very high level, with deep understanding of types, fields, and methods in the program. This abstraction makes it easy for different .NET languages to interoperate, and it also makes it possible to execute the same program binary on different physical processors, with different versions of the program's dependencies.
+
+One of the most important tasks a .NET runtime has is turning instructions for the abstract machine into actions on a concrete hardware. There are many ways to do this.
+
+### Just in time compilation
+
+One of the techniques that CoreCLR employs is Just in time compilation ("JIT"). This strategy translates instructions for the abstract processor into native instructions for the processor that the program is running on "just in time". The term "just in time" means that the translation happens when the need arises. For example, a method can be translated when it's first called.
+
+The actual transformation of IL into native code is handled by the code generator. Code generator is a component of the CoreCLR virtual machine that (with the help of other components of CoreCLR, such as the type system) translates IL into native code. The code generator talks to the rest of the virtual machine over a well-defined interface: this allows code generators to be relatively pluggable. The code generator used by the CoreCLR is [RyuJIT](../botr/ryujit-overview.md). Over the years, CLR has had many other code generators serving different purposes, such as the simplified [fjit](https://github.com/SSCLI/sscli20_20060311/tree/master/clr/src/fjit), LLVM-based [LLIC](https://github.com/dotnet/llilc), or the closed-source jit32 and jit64.
+
+Big advantage of Just in time compilation is that the generated native code can be tailored for the specific physical processor model. RyuJIT currently uses information about the processor to e.g. unlock the use of AVX2 instructions on x64 processors.
+
+Just in time compilation needs to carefully balance the time it takes to compile the method with the level of code generation optimizations performed.
```

Comment:
Maybe add a comment on Tiered Compilation here and link to the section further down. Just something that calls it out as a technique to help maintain that balance?

Reply:
I find it more readable if a document starts with simple things and then increases difficulty by adding more concepts as I read it.

Adding more concepts that are not essential in the beginning tends to detract from the main points. Reference to tiered compilation here would only be useful if the reader already knows what tiered compilation is. Otherwise they're going to jump around the doc.